06-security.md

```diff
@@ -1 +1,426 @@
+
+The session id must change after a successful login and logout. Otherwise session fixation attacks may occur.  
+
+### Use Secure Session Tokens
+
+Session Cookies must have the flags `http-only` and `secure` set. Http-only means that the cookie is not accessible from Javascript (otherwise the session could be stolen using cross-site-scripting). Secure means that it can only be transmitted over a HTTPS connection (otherwise the session could be read in transit by a man-in-the-middle).
+
+### Protect the Session against Cross Site Request Forgery (XSRF)
+
+A malicious browser tab may send requests to the API while the user is logged in. These requests will be sent with the correct session id by the browser. 
+To prevent abuse an additional token must be sent with each write operation, the `XSRF-TOKEN`. The token name used in this convention is the default for AngularJS.  
+
+Typically the token is sent to the browser in the form of a cookie by the first GET request to the API, read by the JavaScript on the page and added as a request header (`X-XSRF-TOKEN`) to subsequent PUT, POST or DELETE request.  The API checks the presence of the correct header on the request. If not correct, the request fails.
+ 
+The malicious tab cannot read the contents of the `XSRF-TOKEN` cookie as it is on a different domain. 
+    
+An XSRF token is assigned to the browser on the first GET request. XSRF cookies MUST have the flag `secure` set. The API must check PUT, POST, PATCH and DELETE requests for the presence of the correct XSRF token.
+
+### Authentication
+
+Usernames and passwords are still the most commonly used way to authenticate a user especially if there is no supporting infrastructure already in place. However implementing proper authentication using usernames and passwords is not a trivial task. Owasp has a [whole page](https://www.owasp.org/index.php/Authentication_Cheat_Sheet) on it. This convention describes just the minimal requirements for authenticating using a username and password. Other (important) subjects such as 'Forgotten Password' functionality or account activation is out of scope. 
+
+#### Passwords must be salted and hashed using BCrypt
+
+BCrypt is a computational heavy hashing algorithm with built in salt. If the user database is stolen it will take a lot more effort to brute force the passwords than by using conventional algorithms (MD5, SHA-1, SHA-256). Note that there is a better hashing algorithm called Argon2 but this is not (yet) supported by Spring-Security.  
+
+#### A failed login attempt must always return the same response.
+
+When authentication fails the username may not exist, the password may be incorrect or the account may have been locked. If an error message is returned it should always be a generic error message ('Authentication Failed') that does not state the reason for failing. This prevents attackers from learning if actually an account exists with that username.
```

Comment:
Maybe it's also a good idea to (briefly) note that the same rule should apply to the 'Forgot Password' functionality. If that _does_ indicate whether or not a username is known then they also have that information.